Implement Zenodo release automation in GitHub Actions

Prateekshit73 opened the pull request:
## Proposed change

Implement Zenodo release automation in GitHub Actions
Closes #2305

## Type of change

- [ ] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [x] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/pin/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new `holidays` functionality in general)

## Checklist


- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've successfully run `make check`, all checks and tests are green

[contributing-guidelines]: https://github.com/vacanza/holidays/blob/dev/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/holidays/tree/dev/docs/source

arkid15r reviewed (changes requested):
Thanks for working on this @Prateekshit73 
This implementation is a bit far from what I meant in the issue description. Specifically the metadata control flexibility

I suggest using https://github.com/megasanjay/upload-to-zenodo?tab=readme-ov-file#inputs approach (or the entire action) 

It also seems that PR contains unrelated changes that may impact other CI/CD jobs.

Prateekshit73 commented:
Thank you for your feedback, and I apologize for not aligning my implementation with the issue description, particularly regarding the metadata control flexibility. I appreciate you pointing this out.

I will take a closer look at the approach suggested in the GitHub link you provided and make the necessary adjustments to ensure it meets the requirements. I also understand the importance of keeping the PR focused, so I will review it for any unrelated changes that could impact other CI/CD jobs and correct those as well.

Thank you for your guidance, and I’ll work on these changes promptly.

KJhellico commented on .github/workflows/ci-cd.yml:
```diff
@@ -311,3 +311,69 @@ jobs:
+        run: |
+          zip -r vacanza-holidays-${{ github.ref_name }}.zip dist/* holidays-*-sbom.json
+
+      - name: Create new Zenodo deposition
+        id: create_deposition
+        run: |
+          response=$(curl -X POST \
+            -H "Authorization: Bearer ${{ secrets.ZENODO_TOKEN }}" \
+            -H "Content-Type: application/json" \
+            -d '{"metadata": {"title": "Vacanza Holidays ${{ github.ref_name }}", "upload_type": "software", "description": "Release ${{ github.ref_name }} of Vacanza Holidays"}}' \
+            https://zenodo.org/api/deposit/depositions)
+          if [[ $(echo "$response" | jq -r '.status // "none"') == "error" ]]; then
+            echo "Error creating deposition: $(echo "$response" | jq -r '.message')"
+            exit 1
+          fi
+          deposition_id=$(echo "$response" | jq -r '.id')
+          echo "DEPOSITION_ID=$deposition_id" >> $GITHUB_ENV
+        env:
+          ZENODO_TOKEN: ${{ secrets.ZENODO_TOKEN }}
```
As I noted [here](https://github.com/vacanza/holidays/pull/2317#issue-2885504518), I recommend to add a new version to the existing Zenodo record instead of creating a new one.

arkid15r commented on .github/workflows/ci-cd.yml:
```diff
@@ -311,3 +311,118 @@ jobs:
+        run: |
+          zip -r vacanza-holidays-${{ github.ref_name }}.zip dist/* holidays-*-sbom.json
+
+      - name: Check for existing Zenodo deposition
```
It looks like we should create our own action for this, it seems https://docs.github.com/en/actions/sharing-automations/creating-actions/creating-a-composite-action would fit.

If possible let's have it in Python

arkid15r reviewed (changes requested):
@Prateekshit73 what's the next step on this task, any blockers?

Prateekshit73 commented:
Hi @arkid15r, thanks for checking in! I’ve been testing some things on the Zenodo sandbox to get a better feel for the deposition process and ensure the workflow behaves as expected. No major blockers so far—just ironing out a few details to make sure it’s robust.

For the next steps, I’ll focus on addressing the feedback from you. I’ve noted your suggestion about creating a composite action in Python—I think that’s a great idea to encapsulate this logic cleanly. I’ll start drafting that in a separate action within the repo, following the composite action guidelines you linked. Should I create it as a new file under `.github/actions/` and reference it in the `ci-cd.yml` workflow?

Let me know if there’s anything specific you’d like me to prioritize or tweak as I move forward. Appreciating the guidance!

arkid15r commented:
>Should I create it as a new file under `.github/actions/` and reference it in the `ci-cd.yml` workflow?
> 
> Let me know if there’s anything specific you’d like me to prioritize or tweak as I move forward. Appreciating the guidance!

That's an option for sure. I think we can afford to create a separate repository for this action, something like https://github.com/vacanza/zenodo-upload (I'm open for better naming options)

KJhellico commented on .github/actions/zenodo-upload/action.yml:
```diff
@@ -0,0 +1,37 @@
+    value: ${{ steps.upload.outputs.doi }}
+runs:
+  using: 'composite'
+  steps:
```
We need to install Python first.

arkid15r commented:
Just FYI -- I created https://github.com/vacanza/update-zenodo-action

Prateekshit73 commented:
Hi @arkid15r,

Thanks for the update and for creating the vacanza/update-zenodo-action repository—love the reusability angle! Given this new setup, would you prefer I close this PR and start a fresh one in vacanza/update-zenodo-action to handle the integration, or should I keep refining this one to use the new action? Also, if there’s anything specific you’d like me to test or focus on with this transition, just let me know.

I have been studying some aspects related to the implementation and want to present a solid logic from my side, which is why it’s taking a bit longer. I apologize for the delay, and I really appreciate your guidance!

Looking forward to your thoughts!

arkid15r commented:
> Hi @arkid15r,
> 
> Thanks for the update and for creating the vacanza/update-zenodo-action repository—love the reusability angle! Given this new setup, would you prefer I close this PR and start a fresh one in vacanza/update-zenodo-action to handle the integration, or should I keep refining this one to use the new action? Also, if there’s anything specific you’d like me to test or focus on with this transition, just let me know.
> 
> I have been studying some aspects related to the implementation and want to present a solid logic from my side, which is why it’s taking a bit longer. I apologize for the delay, and I really appreciate your guidance!
> 
> Looking forward to your thoughts!

Yes, please use the new repository for your code. I suggest focusing on making the action as reusable as possible (not only by us -- by external users too).

Prateekshit73 commented:
Closing this PR as per @arkid15r's suggestion to move work to `vacanza/update-zenodo-action`

Prateekshit73 commented:
Thanks for the direction, @arkid15r ! I’ll start working in `vacanza/update-zenodo-action`. I’ll focus on reusability as suggested and keep you posted on progress. Let me know if there’s anything specific you’d like me to prioritize!
